File: python/examples/text_editor/text_editor.py
"""Simple local text editor example."""

import os
import sys

_here = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.dirname(os.path.dirname(_here)))

from runtime import UIApplication
from dialogs import FileBrowser, MessageBoxYesNoCancel
from text_editor_ui import TextEditorUI


NOTES_DIR_ENV = "UI_TEXT_EDITOR_NOTES_DIR"
DEFAULT_NOTE_ENV = "UI_TEXT_EDITOR_DEFAULT_NOTE"
BROWSER_ROOT_DIR_ENV = "UI_TEXT_EDITOR_BROWSER_ROOT_DIR"
DEFAULT_BROWSER_DIR_ENV = "UI_TEXT_EDITOR_BROWSER_DIR"
NOTES_DIR = os.environ.get(NOTES_DIR_ENV, os.path.join(_here, "notes"))
DEFAULT_NOTE = os.environ.get(DEFAULT_NOTE_ENV, "welcome.txt")
UNTITLED_NAME = "untitled.txt"
USER_HOME_DIR = os.path.expanduser("~")
DESKTOP_DIR = os.path.join(USER_HOME_DIR, "Desktop")
BROWSER_ROOT_DIR = os.environ.get(BROWSER_ROOT_DIR_ENV, USER_HOME_DIR)
DEFAULT_BROWSER_DIR = os.environ.get(
    DEFAULT_BROWSER_DIR_ENV,
    DESKTOP_DIR if os.path.isdir(DESKTOP_DIR) else USER_HOME_DIR,
)


class TextEditor(TextEditorUI):
    def open(self):
        super().open()
        self._loaded_text = ""
        self._current_path = os.path.join(NOTES_DIR, DEFAULT_NOTE)
        self._pending_action = None
        self._after_save_action = None
        self._load_path(self._current_path)

    def on_new_btn_click(self):
        self._with_saved_changes(self._new_file)

    def on_open_btn_click(self):
        self._with_saved_changes(self._open_browser)

    def on_save_btn_click(self):
        self._save_or_save_as()

    def on_save_as_btn_click(self):
        self._save_as_browser()

    def on_quit_btn_click(self):
        self._with_saved_changes(self._quit)

    def on_editor_change(self, value):
        self._refresh_status("Modified")

    def _new_file(self):
        self._exit_edit_mode(commit=True, notify=False)
        self._current_path = None
        self.filename.text = UNTITLED_NAME
        self.editor.value = ""
        self.editor.cursor_pos = 0
        self.editor._scroll_offset = 0
        self._loaded_text = ""
        self._refresh_status("New file")

    def _load_path(self, path):
        self._exit_edit_mode(commit=True, notify=False)
        path = os.path.abspath(path)
        try:
            with open(path, "r", encoding="utf-8") as handle:
                text = handle.read()
        except OSError:
            text = ""
        self._current_path = path
        self.filename.text = os.path.basename(path)
        self.editor.value = text
        self.editor.cursor_pos = 0
        self.editor._scroll_offset = 0
        self._loaded_text = text
        self._refresh_status("Loaded")

    def _save_or_save_as(self):
        if self._current_path is None:
            self._save_as_browser()
        else:
            self._save_path(self._current_path)

    def _save_path(self, path):
        path = os.path.abspath(path)
        with open(path, "w", encoding="utf-8") as handle:
            handle.write(self.editor.value)
        self._current_path = path
        self.filename.text = os.path.basename(path)
        self._loaded_text = self.editor.value
        self._refresh_status("Saved")

    def _open_browser(self):
        browser = FileBrowser(BROWSER_ROOT_DIR, DEFAULT_BROWSER_DIR, mode="open", on_close=self._on_open_selected)
        self._app.open(browser)

    def _save_as_browser(self):
        start = os.path.dirname(self._current_path) if self._current_path else NOTES_DIR
        if not os.path.abspath(start).startswith(os.path.abspath(BROWSER_ROOT_DIR)):
            start = DEFAULT_BROWSER_DIR
        initial_filename = os.path.basename(self._current_path) if self._current_path else UNTITLED_NAME
        browser = FileBrowser(
            BROWSER_ROOT_DIR,
            start,
            mode="save",
            on_close=self._on_save_selected,
            initial_filename=initial_filename,
        )
        self._app.open(browser)

    def _on_open_selected(self, path):
        if path:
            self._load_path(path)

    def _on_save_selected(self, path):
        action = self._after_save_action
        self._after_save_action = None
        if path:
            self._save_path(path)
            if action is not None:
                action()
        elif action is not None:
            self._refresh_status("Cancelled")

    def _with_saved_changes(self, action):
        self._after_save_action = None
        if not self._is_modified():
            action()
            return
        self._pending_action = action
        dialog = MessageBoxYesNoCancel(
            header="Unsaved Changes",
            message="Save current file before continuing?",
            on_close=self._on_unsaved_choice,
        )
        self._app.open(dialog)

    def _on_unsaved_choice(self, choice):
        action = self._pending_action
        self._pending_action = None
        if choice == "cancel" or action is None:
            self._refresh_status("Cancelled")
            return
        if choice == "yes":
            if self._current_path is None:
                self._after_save_action = action
                self._save_as_browser()
                return
            self._save_or_save_as()
            action()
            return
        if choice == "no":
            action()

    def _quit(self):
        if self._app is not None:
            self._app._running = False

    def _refresh_status(self, prefix):
        text = self.editor.value
        modified = "*" if self._is_modified() else ""
        line_count = text.count("\n") + 1 if text else 0
        self.status.text = f"{prefix}: {self.filename.text}{modified} - {line_count} lines, {len(text)} chars"

    def _is_modified(self):
        return self.editor.value != self._loaded_text


def main():
    app = UIApplication()
    window = TextEditor()
    app.open(window)
    return app.run()


if __name__ == "__main__":
    sys.exit(main())

File: java/examples/text_editor/TextEditor.java
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

import uimd.FileBrowser;
import uimd.GeneratedWindowRuntime;
import uimd.MessageBoxYesNoCancel;

public final class TextEditor extends TextEditorUI
{
    private static final String NOTES_DIR_ENV = "UI_TEXT_EDITOR_NOTES_DIR";
    private static final String DEFAULT_NOTE_ENV = "UI_TEXT_EDITOR_DEFAULT_NOTE";
    private static final String BROWSER_ROOT_DIR_ENV = "UI_TEXT_EDITOR_BROWSER_ROOT_DIR";
    private static final String DEFAULT_BROWSER_DIR_ENV = "UI_TEXT_EDITOR_BROWSER_DIR";
    private static final String DEFAULT_FILE_NAME = "welcome.txt";
    private static final String UNTITLED_NAME = "untitled.txt";
    private static final int EXPECTED_LAYOUT_ENTRIES = 10;

    private Path currentPath;
    private String loadedText = "";
    private Runnable pendingAction;
    private Runnable afterSaveAction;
    private boolean closeRequested;

    public TextEditor()
    {
        loadPath(notesDir().resolve(defaultNoteName()));
    }

    @Override
    protected void onNewBtnClick()
    {
        withSavedChanges(this::newFile);
    }

    @Override
    protected void onOpenBtnClick()
    {
        withSavedChanges(this::openBrowser);
    }

    @Override
    protected void onSaveBtnClick()
    {
        saveOrSaveAs();
    }

    @Override
    protected void onSaveAsBtnClick()
    {
        saveAsBrowser();
    }

    @Override
    protected void onQuitBtnClick()
    {
        withSavedChanges(this::quit);
    }

    @Override
    protected void onEditorChange(String value)
    {
        markModified();
    }

    @Override
    protected boolean shouldClose()
    {
        return closeRequested;
    }

    public void newFile()
    {
        currentPath = null;
        filename.setText(UNTITLED_NAME);
        setEditorText("");
        loadedText = "";
        refreshStatus("New file");
    }

    public void loadPath(Path path)
    {
        Path absolutePath = path.toAbsolutePath().normalize();
        String text = readTextFile(absolutePath);
        currentPath = absolutePath;
        filename.setText(displayFileName(absolutePath));
        setEditorText(text);
        loadedText = text;
        refreshStatus("Loaded");
    }

    public void saveOrSaveAs()
    {
        if (currentPath == null)
        {
            saveAsBrowser();
            return;
        }
        savePath(currentPath);
    }

    public void openBrowser()
    {
        FileBrowser browser = new FileBrowser(
            browserRootDir(),
            defaultBrowserDir(),
            FileBrowser.MODE_OPEN,
            this::onOpenSelected,
            "",
            "");
        openWindow(browser, browser.runtimeOptions());
    }

    public void saveAsBrowser()
    {
        Path start = currentPath == null || currentPath.getParent() == null
            ? notesDir()
            : currentPath.getParent();
        if (!pathStartsWith(start, browserRootDir()))
        {
            start = defaultBrowserDir();
        }
        String initialFilename = currentPath == null
            ? UNTITLED_NAME
            : displayFileName(currentPath);
        FileBrowser browser = new FileBrowser(
            browserRootDir(),
            start,
            FileBrowser.MODE_SAVE,
            this::onSaveSelected,
            initialFilename,
            "");
        openWindow(browser, browser.runtimeOptions());
    }

    public void savePath(Path path)
    {
        Path absolutePath = path.toAbsolutePath().normalize();
        writeTextFile(absolutePath, editor.value());
        currentPath = absolutePath;
        filename.setText(displayFileName(absolutePath));
        loadedText = editor.value();
        refreshStatus("Saved");
    }

    public void markModified()
    {
        refreshStatus("Modified");
    }

    public boolean modified()
    {
        return !editor.value().equals(loadedText);
    }

    public void withSavedChanges(Runnable action)
    {
        withSavedChanges(action, "");
    }

    public void withSavedChanges(Runnable action, String simulatedChoice)
    {
        afterSaveAction = null;
        if (!modified())
        {
            action.run();
            return;
        }
        pendingAction = action;
        if (simulatedChoice != null && !simulatedChoice.isEmpty())
        {
            onUnsavedChoice(simulatedChoice);
            return;
        }
        MessageBoxYesNoCancel dialog = new MessageBoxYesNoCancel(
            "Unsaved Changes",
            "Save current file before continuing?",
            this::onUnsavedChoice);
        openWindow(dialog, dialog.runtimeOptions());
    }

    public void quit()
    {
        closeRequested = true;
    }

    public void onOpenSelected(Path path)
    {
        if (path != null)
        {
            loadPath(path);
        }
    }

    public void onSaveSelected(Path path)
    {
        Runnable action = afterSaveAction;
        afterSaveAction = null;
        if (path != null)
        {
            savePath(path);
            if (action != null)
            {
                action.run();
            }
        }
        else if (action != null)
        {
            refreshStatus("Cancelled");
        }
    }

    private void onUnsavedChoice(String choice)
    {
        Runnable action = pendingAction;
        pendingAction = null;
        if ("cancel".equals(choice) || action == null)
        {
            refreshStatus("Cancelled");
            return;
        }
        if ("yes".equals(choice))
        {
            if (currentPath == null)
            {
                afterSaveAction = action;
                saveAsBrowser();
                return;
            }
            saveOrSaveAs();
            action.run();
            return;
        }
        if ("no".equals(choice))
        {
            action.run();
        }
    }

    private void setEditorText(String text)
    {
        editor.setValue(text);
        editor.selectRange(0, 0);
    }

    private void refreshStatus(String prefix)
    {
        String text = editor.value();
        String modifiedMarker = modified() ? "*" : "";
        int lineCount = text.isEmpty() ? 0 : countNewlines(text) + 1;
        status.setText(prefix + ": " + filename.text() + modifiedMarker
            + " - " + lineCount + " lines, " + text.length() + " chars");
    }

    private static int countNewlines(String text)
    {
        int count = 0;
        for (int index = 0; index < text.length(); ++index)
        {
            if (text.charAt(index) == '\n')
            {
                ++count;
            }
        }
        return count;
    }

    private static String environmentValue(String name)
    {
        String value = System.getenv(name);
        return value == null || value.isEmpty() ? null : value;
    }

    private static Path environmentPath(String name)
    {
        String value = environmentValue(name);
        return value == null ? null : Path.of(value).toAbsolutePath().normalize();
    }

    private static Path notesDir()
    {
        Path configured = environmentPath(NOTES_DIR_ENV);
        if (configured != null)
        {
            return configured;
        }
        Path location = classLocation();
        while (location != null)
        {
            Path candidate = location.resolve("notes");
            if (Files.isDirectory(candidate))
            {
                return candidate;
            }
            location = location.getParent();
        }
        return Path.of("notes").toAbsolutePath().normalize();
    }

    private static Path classLocation()
    {
        try
        {
            URI location = TextEditor.class.getProtectionDomain()
                .getCodeSource()
                .getLocation()
                .toURI();
            Path path = Path.of(location).toAbsolutePath().normalize();
            return Files.isDirectory(path) ? path : path.getParent();
        }
        catch (NullPointerException | SecurityException | URISyntaxException exception)
        {
            return Path.of("").toAbsolutePath().normalize();
        }
    }

    private static String defaultNoteName()
    {
        String configured = environmentValue(DEFAULT_NOTE_ENV);
        return configured == null ? DEFAULT_FILE_NAME : configured;
    }

    private static Path userHomeDir()
    {
        String home = environmentValue("HOME");
        if (home == null)
        {
            home = System.getProperty("user.home", System.getProperty("user.dir", "."));
        }
        return Path.of(home).toAbsolutePath().normalize();
    }

    private static Path browserRootDir()
    {
        Path configured = environmentPath(BROWSER_ROOT_DIR_ENV);
        return configured == null ? userHomeDir() : configured;
    }

    private static Path defaultBrowserDir()
    {
        Path configured = environmentPath(DEFAULT_BROWSER_DIR_ENV);
        if (configured != null)
        {
            return configured;
        }
        Path desktop = userHomeDir().resolve("Desktop");
        return Files.isDirectory(desktop) ? desktop : userHomeDir();
    }

    private static boolean pathStartsWith(Path path, Path root)
    {
        return path.toAbsolutePath().normalize().startsWith(root.toAbsolutePath().normalize());
    }

    private static String readTextFile(Path path)
    {
        try
        {
            return Files.readString(path, StandardCharsets.UTF_8);
        }
        catch (IOException exception)
        {
            return "";
        }
    }

    private static void writeTextFile(Path path, String text)
    {
        try
        {
            Files.writeString(path, text, StandardCharsets.UTF_8);
        }
        catch (IOException exception)
        {
            throw new IllegalStateException("failed to save text file: " + path, exception);
        }
    }

    private static String displayFileName(Path path)
    {
        Path name = path.getFileName();
        return name == null ? path.toString() : name.toString();
    }

    public static void main(String[] arguments)
    {
        int exitCode = GeneratedWindowRuntime.runGeneratedAppMain(() -> run(arguments));
        System.exit(exitCode);
    }

    private static int run(String[] arguments)
    {
        if (arguments.length > 0 && "--smoke".equals(arguments[0]))
        {
            smoke();
            return 0;
        }
        if (arguments.length > 0 && "--logic-test".equals(arguments[0]))
        {
            logicTest();
            return 0;
        }
        TextEditor application = new TextEditor();
        return GeneratedWindowRuntime.runGeneratedWindow(
            application,
            application.runtimeOptions(),
            arguments);
    }

    private static void smoke()
    {
        TextEditor window = new TextEditor();
        require("Text Editor".equals(window.title()), "unexpected title");
        require(window.generatedLayout().size() == EXPECTED_LAYOUT_ENTRIES,
            "unexpected layout size");
        require(DEFAULT_FILE_NAME.equals(window.filename.text()), "unexpected default filename");
    }

    private static void logicTest()
    {
        TextEditor window = new TextEditor();
        window.newFile();
        require(UNTITLED_NAME.equals(window.filename.text()), "new filename mismatch");
        require(window.editor.value().isEmpty(), "new editor should be empty");
        window.editor.setValue("draft");
        window.markModified();
        require(window.modified(), "draft should be modified");
        require("Modified: untitled.txt* - 1 lines, 5 chars".equals(window.status.text()),
            "modified status mismatch");
        boolean[] actionCalled = {false};
        window.withSavedChanges(() -> actionCalled[0] = true, "no");
        require(actionCalled[0], "discard action was not called");

        Path savedPath = null;
        try
        {
            savedPath = Files.createTempFile("uimd-java-text-editor-", ".txt");
            window.editor.setValue("saved draft");
            window.onSaveSelected(savedPath);
            require("saved draft".equals(readTextFile(savedPath)), "saved content mismatch");
            require(!window.modified(), "saved document should not be modified");
        }
        catch (IOException exception)
        {
            throw new IllegalStateException("failed to create text editor test file", exception);
        }
        finally
        {
            if (savedPath != null)
            {
                try
                {
                    Files.deleteIfExists(savedPath);
                }
                catch (IOException exception)
                {
                    throw new IllegalStateException("failed to remove text editor test file", exception);
                }
            }
        }
    }

    private static void require(boolean condition, String message)
    {
        if (!condition)
        {
            throw new IllegalStateException(message);
        }
    }
}
